beluga/include/beluga/views/low_variance_sample.hpp

```diff
@@ -0,0 +1,223 @@
+/// and Autonomous Agents). The MIT Press. https://gaoyichao.com/Xiaotu/resource/refs/PR.MIT.en.pdf
+
+/**
+ * \tparam Range A [random access](https://en.cppreference.com/w/cpp/ranges/random_access_range) and
+ *  [sized](https://en.cppreference.com/w/cpp/ranges/sized_range) range.
+ * \tparam Weights
+ * \tparam URNG A random number generator that satisfies the
+ *  [UniformRandomBitGenerator](https://en.cppreference.com/w/cpp/named_req/UniformRandomBitGenerator)
+ *  requirements.
+ */
+template <class Range, class Weights, class URNG = typename ranges::detail::default_random_engine>
+struct low_variance_sample_view
+    : public ranges::view_facade<low_variance_sample_view<Range, Weights, URNG>, ranges::infinite> {
+ public:
+  /// Default constructor.
+  low_variance_sample_view() = default;
+
+  /// Construct the view from an existing range.
+  /**
+   * \param range The range to be adapted.
+   * \param weights The weights associated with the elements in the range.
+   * \param engine The random number generator object.
+   */
+  constexpr low_variance_sample_view(Range range, Weights weights, URNG& engine = ranges::detail::get_random_engine())
+      : range_{std::move(range)}, weights_{std::move(weights)}, engine_{std::addressof(engine)} {
+    assert(ranges::size(range) > 0);
+  }
+
+ private:
+  // `ranges::range_access` needs access to the cursor members.
+  friend ranges::range_access;
+
+  static_assert(ranges::sized_range<Range>);
+  static_assert(ranges::random_access_range<Range>);
+  static_assert(ranges::input_range<Weights>);
+
+  /// Cursor class that handles the iteration logic.
+  struct cursor {
+   public:
+    /// Default constructor.
+    cursor() = default;
+
+    /// Construct a cursor from the parent view elements.
+    constexpr explicit cursor(low_variance_sample_view* view)
+        : view_(view),
+          range_begin_{ranges::begin(view_->range_)},
+          weights_begin_{ranges::begin(view_->weights_)},
+          M_{static_cast<long long>(ranges::size(view_->range_))},
+          r_{std::uniform_real_distribution<double>(0.0, 1.0 / static_cast<double>(M_))(
+              *view_->engine_)},  // Generate r only once
+          m_{0},
+          i_{0},
+          c_{*weights_begin_} {}
+
+    /// Access the current iterator.
+    [[nodiscard]] constexpr decltype(auto) read() const noexcept(noexcept(*range_begin_)) { return *it_; }
+
+    /// Position the current iterator.
+    constexpr void next() {
+      ++m_;
+      const double U = r_ + (static_cast<double>(m_) - 1.0) / static_cast<double>(M_);
+      while (U > c_) {
+        ++i_;
+        c_ += static_cast<double>(*std::next(weights_begin_, i_));  // weights_begin_[i_]
+      }
+      it_ = std::next(range_begin_, i_);  // range_begin_ + i_
+    }
+
+    /// Returns true if the cursor is at the end of the range.
+    constexpr bool equal(const cursor& other) const noexcept { return m_ == other.M_; }
```

Comment:
@larodriguez22 hmm, we are using a sentinel as end of iteration, why do we need this?